common/rate-limiter/integration-test/src/main.rs

```diff
@@ -0,0 +1,213 @@
+
+        let mut listener = tokio::net::TcpListener::bind(ADDRESS)
+            .await
+            .expect("Must bind to address");
+        eprintln!("Connecting to {}", ADDRESS);
+        let mut writer = writer(
+            tokio::net::TcpStream::connect(ADDRESS)
+                .await
+                .expect("Must connect to address"),
+        );
+
+        let (stats_sender, mut stats_receiver) = channel(2);
+        let mut read_stats = stats_sender.clone();
+        let mut write_stats = stats_sender.clone();
+
+        // Ensure we're listening first
+        let read_future = async move {
+            let in_stream = listener
+                .next()
+                .await
+                .expect("Should not have an error connecting")
+                .expect("Should have an incoming connection");
+            let mut reader = reader(in_stream);
+            let stats = stats("Read", || async {
+                let mut buf: [u8; CHUNK_SIZE] = [0; CHUNK_SIZE];
+                let mut num_reads: usize = 0;
+                let mut num_bytes: usize = 0;
+                loop {
+                    num_bytes += reader.read(&mut buf).await.unwrap();
+                    num_reads += 1;
+                    if num_bytes >= total_test_bytes {
+                        break;
+                    }
+                }
+                (num_reads, num_bytes)
+            })
+            .await;
+
+            read_stats.send(stats).await.unwrap();
+        };
+
+        let write_future = async move {
+            let stats = stats("Write", || async {
+                let source: [u8; CHUNK_SIZE] = [1; CHUNK_SIZE];
+                let mut num_writes: usize = 0;
+                let mut num_bytes: usize = 0;
+                loop {
+                    num_bytes += writer.write(&source).await.unwrap();
+                    num_writes += 1;
+                    if num_bytes >= total_test_bytes {
+                        break;
+                    }
+                }
+                (num_writes, num_bytes)
+            })
+            .await;
+            write_stats.send(stats).await.unwrap();
+        };
+
+        // Wait for both to finish
+        futures::future::join(read_future, write_future).await;
+
+        let mut results: Vec<RunStats> = vec![];
+        // Print out write and read stats
+        for _ in 0u8..2u8 {
+            results.push(stats_receiver.next().await.unwrap());
+        }
+
+        results
+    }
+
+    fn simple_shared_bucket(size: usize) -> SharedBucket {
+        Arc::new(Mutex::new(Bucket::new(size, size, size)))
+    }
+
+    /// Tests backpressure, where only the reader is limited.  Using a socket adds a buffer
+    /// that causes the backpressure only to occur after a large amount of traffic first. (> 500k bytes)
+    /// > 55 chunks could cause backpressure, for a proper test > 100 or > 200 give good results.
+    pub async fn test_rate_limit_read(num_chunks: usize) -> Vec<RunStats> {
```

Comment:
would recommend replacing num_chunks with num_bytes as chunks requires digging into the code to understand what's going on.